src/amb_kf.c

```diff
@@ -38,75 +43,102 @@
 
 
-  double g[state_dim]; /*  g = diag(D) * f. */
-  double alpha = R;    /*  alpha = f * g + R = f^T * diag(D) * f + R. */
+  /*  g = diag(D) * f.
+      alpha = f * g + R = f^T * diag(D) * f + R. */
+  double alpha = R;
   for (u32 i=0; i<state_dim; i++) {
     g[i] = D[i] * f[i];
     alpha += f[i] * g[i];
   }
-  if (DEBUG) {
-    VEC_PRINTF(f, state_dim);
-    VEC_PRINTF(g, state_dim);
-    printf("alpha = %.16f", alpha);
-    if (abs(alpha) == 0) {
-      printf(" \t (alpha == 0 exactly)\n");
-    }
-    else {
-      printf("\n");
-    }
+  return alpha;
+}
+
+
+/** In place updating of the state cov and k vec using a scalar observation
+ * This is from section 10.2.1 of Gibbs [1], with some extra logic for handling
+ * singular matrices, dictating that zeros from cov_D dominate in a particular
+ * potential 0 / 0.
+ * We also make it more robust, by multiplying k by k_scalar <=  1.
+ * To still perform the update in factored form, we modified the derivation
+ * in section V.3 of [2] to arrive at the following code.
+ *
+ * \param kf        The KF to update
+ * \param R         The measurement variance
+ * \param f         U^T * h
+ * \param g         diag(D) * f
+ * \param alpha     The innovation variance
+ * \param k_scalar  A scalar to multiply the Kalman gain by (softens outliers)
+ * \param innov     The difference between the actual and predicted observation
+ */
+static void update_kf_state(nkf_t *kf, double R, double *f, double *g,
+                            double alpha, double k_scalar,
+                            double innov)
+{
+  DEBUG_ENTRY();
+
+  assert(k_scalar <= 1);
+  assert(k_scalar >= 0);
+  if (k_scalar == 0) {
+    return;
   }
+  u32 state_dim = kf->state_dim;
+  double *U = kf->state_cov_U;
+  double *D = kf->state_cov_D;
+  double k[state_dim];
 
-  double gamma[state_dim];
   double U_bar[state_dim * state_dim];
   double D_bar[state_dim];
 
-  memset(gamma, 0,             state_dim * sizeof(double));
   memset(U_bar, 0, state_dim * state_dim * sizeof(double));
   memset(D_bar, 0,             state_dim * sizeof(double));
   memset(k,     0,             state_dim * sizeof(double));
 
-  gamma[0] = R + g[0] * f[0];
+  /* Turns out that to use k' = k * k_scalar instead of k as the Kalman gain,
+   * the only change in the covariance calculation needed is the initialization
+   * of gamma (alpha_i in Bierman's old book [2] V.3) */
+  double gamma = R / k_scalar + g[0] * f[0];
+  if (k_scalar < 1) {
+    for (u8 i = 0; i < kf->state_dim; i++) {
+      gamma += (1 - k_scalar) / k_scalar * g[i] * f[i];
+    }
+  }
   if (D[0] == 0 || R == 0) {
     /*  This is just an expansion of the other branch with the proper
      *  0 `div` 0 definitions. */
     D_bar[0] = 0;
   }
   else {
-    D_bar[0] = D[0] * R / gamma[0];
+    D_bar[0] = D[0] * R / gamma;
   }
   k[0] = g[0];
   U_bar[0] = 1;
   if (DEBUG) {
-    printf("gamma[0] = %f\n", gamma[0]);
+    printf("gamma[0] = %f\n", gamma);
```

Comment:
You know my preference is to get rid of these `DEBUG` sections cluttering the code :) Shouldn't be so necessary if the code its thoroughly unit tested.